File: Rust/2015/day01/src/main.rs
#![allow(unused)]

use std::io::{self, Read};
use std::vec;
use std::io::{Write, BufReader, BufRead, ErrorKind};
use std::fs::File;
use std::cmp::Ordering;
use std::collections::{HashMap, HashSet};

fn read_file(file_name: &str) -> String {
    let mut file = File::open(file_name).unwrap();
    let mut contents = String::new();
    file.read_to_string(&mut contents).unwrap();
    contents
}

fn main() {
    let input = read_file("input.txt");

    let mut floor = 0;
    let mut p2 = false;
    let mut p2_ans = 0;

    for (i, c) in input.chars().enumerate() {
        match c {
            '(' => floor += 1,
            ')' => floor -= 1,
            _ => continue,
        }

        if floor < 0 && !p2 {
            p2_ans = i + 1;
            p2 = true;
        }
    }

    println!("Part 1: {}", floor);
    println!("Part 2: {}", p2_ans);
}

File: Go/2015/day01/main.go
package main

import (
	"fmt"
	"os"
)

func getInput(fileName string) string {
	bytes, err := os.ReadFile(fileName)
	if err != nil {
		panic(err)
	}

	return string(bytes)
}

func part1(input string) {
	floor := 0

	for _, char := range input {
		if char == '(' {
			floor += 1
		} else {
			floor -= 1
		}
	}

	fmt.Printf("Part 1: %d\n", floor)
}

func part2(input string) {
	floor := 0
	counter := 0

	for floor >= 0 {
		if input[counter] == '(' {
			floor += 1
		} else {
			floor -= 1
		}

		counter += 1
	}

	fmt.Printf("Part 2: %d\n", counter)
}

func main() {
	input := getInput("input.txt")
	part1(input)
	part2(input)
}
